assets/js/stripe.js

```diff
@@ -398,305 +508,281 @@ jQuery( function( $ ) {
 			if ( response.error ) {
 				return $( document.body ).trigger( 'stripeError', response );
 			}
 
-			var intent = response.paymentIntent;
-
 			wc_stripe_form.reset();
-			wc_stripe_form.form.append( "<input type='hidden' class='stripe-intent' name='stripe_intent' value='" + intent.id + "'/>" );
+			wc_stripe_form.addHiddenInput( 'source', response.source.id );
 
-			// ToDo: Check for `form#add_payment_method` and remove event listeners
+			if ( $( 'form#add_payment_method' ).length ) {
+				$( wc_stripe_form.form ).off( 'submit', wc_stripe_form.form.onSubmit );
+			}
 
 			wc_stripe_form.form.submit();
 		},
 
-		processStripeResponse: function( source ) {
-			wc_stripe_form.reset();
+		/**
+		 * Responds to payments with credit cards, which use PaymentIntents.
+		 *
+		 * @param {object} response The response from calling `stripe.handleCardPayment`.
+		 */
+		paymentIntentResponse: function( response ) {
+			if ( response.error ) {
+				return $( document.body ).trigger( 'stripeError', response );
+			}
 
-			// Insert the Source into the form so it gets submitted to the server.
-			wc_stripe_form.form.append( "<input type='hidden' class='stripe-source' name='stripe_source' value='" + source.id + "'/>" );
+			wc_stripe_form.reset();
+			wc_stripe_form.addHiddenInput( 'intent', response.paymentIntent.id );
 
-			if ( $( 'form#add_payment_method' ).length ) {
-				$( wc_stripe_form.form ).off( 'submit', wc_stripe_form.form.onSubmit );
-			}
+			// ToDo: Check for `form#add_payment_method` and remove event listeners
 
 			wc_stripe_form.form.submit();
 		},
 
+		/**
+		 * Performs payment-related actions when a checkout/payment form is being submitted.
+		 *
+		 * @param {Event} e A form submission event.
+		 */
 		onSubmit: function( e ) {
 			if ( ! wc_stripe_form.isStripeChosen() ) {
 				return;
 			}
 
-			if ( ! wc_stripe_form.isStripeSaveCardChosen() && ! wc_stripe_form.hasSource() && ! wc_stripe_form.hasToken() && ! wc_stripe_form.hasIntent() ) {
-				e.preventDefault();
+			/**
+			 * ToDo: The logic here was restructured a bit, make sure it still works well.
+			 */
 
-				wc_stripe_form.block();
+			// If a source, token, or an intent is already in place, submit the form as usual.
+			if ( wc_stripe_form.isStripeSaveCardChosen() || wc_stripe_form.hasSource() || wc_stripe_form.hasToken() || wc_stripe_form.hasIntent() ) {
+				return true;
+			}
 
-				// Stripe Checkout.
-				if ( 'yes' === wc_stripe_params.is_stripe_checkout && wc_stripe_form.isStripeModalNeeded() && wc_stripe_form.isStripeCardChosen() ) {
-					if ( 'yes' === wc_stripe_params.is_checkout ) {
-						return true;
-					} else {
-						wc_stripe_form.openModal();
-						return false;
-					}
+			// Open the Stripe Checkout modal.
+			if ( 'yes' === wc_stripe_params.is_stripe_checkout && wc_stripe_form.isStripeModalNeeded() && wc_stripe_form.isStripeCardChosen() ) {
+				if ( 'yes' === wc_stripe_params.is_checkout ) {
+					return true;
```

Comment:
In the previous logic, `e.preventDefault()` was being called even in this scenario. I don't get exactly what this is supposed to accomplish, but I'd just slap a `e.preventDefault()` before this just in case.

Reply:
It's pretty much redundant in combination with the `return`s. There is no difference in both the add card and payment forms, so I think it should be safe to keep it this way and be consistent with the other changes.

Reply:
I still don't get it. For what I understand, this code runs when both:
- The user chose a Stripe payment method (`wc_stripe_form.isStripeCardChosen()`)
- The user is in the checkout page (`'yes' === wc_stripe_params.is_checkout`)
- The merchant is using the "Legacy Checkout" feature (`wc_stripe_form.isStripeModalNeeded()`)

So, that means that the form must be submitted normally, so the user gets to the "Pay for this" page and invoke the Stripe Modal there. So, I understand the `return true;`.

What I don't understand is, that in the old version of this code, `e.preventDefault();` (and `form.block()`) was called before the `return true;` even in this case, so the form wouldn't get submitted.

Does Legacy Checkout work in `master` at all? I don't see how it can work.

Reply:
Took me a while to realise what's going on too, sorry I didn't document it.

<img width="795" alt="screenshot 2019-03-01 at 14 38 17" src="https://user-images.githubusercontent.com/5311119/53641568-b2a81480-3c2f-11e9-8998-e1da36089574.png">

Source: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/master/assets/js/stripe.js#L239  🤣

Reply:
🤦‍♂️ Ok, that makes sense in its own way.